opt: fix logical properties for aggregates which can be NULL

RaduBerinde opened the pull request:
There are a few aggregates which can return NULL even if they have a
non-NULL input: variance, stddev, corr require at least two not-null
input values.

This change fixes the logical props builder to not set the columns as
not-nullable in these cases. It also improves the code around the
functions describing various properties of aggregates to make sure we
didn't miss any aggregates (and we don't miss adding any future new
ones).

Fixes #46412.

Release note (bug fix): fixed incorrect query results in some
cornercases involving variance/stddev/corr.

Release justification: high benefit, low risk change to existing
functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46436)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thanks for fixing

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46436#-:-M38EiSs14BAY2EmYTFx:b-4bx4ne)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ok, apparently it was necessary to go to https://bors.crdb.io/repositories/56/settings and trigger a refresh of the list. Should work now, fingers crossed.

RaduBerinde commented:
Had to make a small fix to `AggsCanBeDecorrelated`. Its behavior should be unchanged.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46436#-:-M38Wdhk8jWL-SvXmuic:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/opt/operator.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46436#-M38TysR06VBFBHo6wNo:-M38TysS1wxXlAy0W4nc:bcymyai) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dab2f1d497108ebbe4f5d8686077234a0da6d03/pkg/sql/opt/operator.go#L277)):*
> ```Go
> 
> 	case AnyNotNullAggOp, AvgOp, BitAndAggOp, BitOrAggOp, BoolAndOp, BoolOrOp,
> 		ConstNotNullAggOp, CountOp, CorrOp, MaxOp, MinOp, SqrDiffOp, StdDevOp,
> ```

Did you intend to make these alphabetical? If so `CorrOp` should come before `CountOp`


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: fix logical properties for aggregates which can be NULL

There are a few aggregates which can return NULL even if they have a
non-NULL input: variance, stddev, corr require at least two not-null
input values.

This change fixes the logical props builder to not set the columns as
not-nullable in these cases. It also improves the code around the
functions describing various properties of aggregates to make sure we
didn't miss any aggregates (and we don't miss adding any future new
ones).

Fixes #46412.

Release note (bug fix): fixed incorrect query results in some
cornercases involving variance/stddev/corr.

Release justification: high benefit, low risk change to existing
functionality.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46436)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/operator.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46436#-M38TysR06VBFBHo6wNo:-M38YyDCErYndaTmuH-T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dab2f1d497108ebbe4f5d8686077234a0da6d03/pkg/sql/opt/operator.go#L277)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Did you intend to make these alphabetical? If so `CorrOp` should come before `CountOp`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46436)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
In 19.2 and earlier, we don't set any aggregation columns as not-null, so the fix doesn't need to be backported.
